File: frontend/src/shared/api/stocks.ts
import { apiRequest } from "./client";
import type { IntradayPoint, LivePriceResponse, PricePoint, Stock } from "../types/stock";

export function fetchStocks(): Promise<Stock[]> {
  return apiRequest<Stock[]>("/api/v1/stocks");
}

export function fetchStockPrices(ticker: string): Promise<PricePoint[]> {
  return apiRequest<PricePoint[]>(`/api/v1/stocks/${ticker}/prices`);
}

export function fetchLivePrice(ticker: string): Promise<LivePriceResponse> {
  return apiRequest<LivePriceResponse>(`/api/v1/stocks/${ticker}/live-price`);
}

export function fetchIntradayPrices(ticker: string): Promise<IntradayPoint[]> {
  return apiRequest<IntradayPoint[]>(`/api/v1/stocks/${ticker}/intraday-prices`);
}

File: backend/app/api/routes/stocks.py
from datetime import datetime

from fastapi import APIRouter, HTTPException

from app.schemas.stock import IntradayPoint, LivePriceResponse, PricePoint, Stock
from app.services import market_data_service

router = APIRouter(prefix="/api/v1/stocks", tags=["stocks"])


@router.get("", response_model=list[Stock])
def list_stocks() -> list[Stock]:
    return market_data_service.get_stocks()


@router.get("/{ticker}/prices", response_model=list[PricePoint])
def get_prices(ticker: str) -> list[PricePoint]:
    if market_data_service.get_stock(ticker) is None:
        raise HTTPException(status_code=404, detail=f"Unknown ticker: {ticker}")
    return market_data_service.get_price_series(ticker)


@router.get("/{ticker}/live-price", response_model=LivePriceResponse)
def get_live_price(ticker: str) -> LivePriceResponse:
    if market_data_service.get_stock(ticker) is None:
        raise HTTPException(status_code=404, detail=f"Unknown ticker: {ticker}")

    live = market_data_service.get_live_price(ticker)
    if live is None:
        return LivePriceResponse(available=False)
    return LivePriceResponse(available=True, as_of=datetime.now().isoformat(), live=live)


@router.get("/{ticker}/intraday-prices", response_model=list[IntradayPoint])
def get_intraday_prices(ticker: str) -> list[IntradayPoint]:
    if market_data_service.get_stock(ticker) is None:
        raise HTTPException(status_code=404, detail=f"Unknown ticker: {ticker}")
    return market_data_service.get_intraday_prices(ticker)
